src/authvalue.rs

```diff
@@ -0,0 +1,101 @@
+use crate::error::TpmError;
+use crypto_glue::{
+    aes256::{self, Aes256Key},
+    argon2, hex,
+    rand::{self, RngCore},
+    zeroize::Zeroizing,
+};
+use std::str::FromStr;
+use tracing::error;
+
+pub enum AuthValue {
+    Key256Bit { auth_key: Aes256Key },
+}
+
+impl AuthValue {
+    fn random_key() -> Result<Zeroizing<[u8; 24]>, TpmError> {
+        let mut auth_key = Zeroizing::new([0; 24]);
+
+        let mut rng = rand::thread_rng();
+        rng.fill_bytes(auth_key.as_mut());
+
+        Ok(auth_key)
+    }
+
+    pub fn generate() -> Result<String, TpmError> {
+        let ak = Self::random_key()?;
+        Ok(hex::encode(&ak))
+    }
+
+    pub fn ephemeral() -> Result<Self, TpmError> {
+        let auth_key = aes256::new_key();
+
+        Ok(AuthValue::Key256Bit { auth_key })
+    }
+
+    /// Derive an auth value from input bytes. This value must be at least 24 bytes in length.
+    ///
+    /// The key derivation is performed with Argon2id.
+    pub fn derive_from_bytes(cleartext: &[u8]) -> Result<Self, TpmError> {
+        use argon2::{Algorithm, Argon2, Params, Version};
+
+        let mut auth_key = Aes256Key::default();
+
+        // This can't be changed else it will break key derivation for users.
+        let argon2id_params =
+            Params::new(32_768, 4, 1, Some(aes256::key_size())).map_err(|argon_err| {
+                error!(?argon_err);
+                TpmError::AuthValueDerivation
+            })?;
+
+        let argon = Argon2::new(Algorithm::Argon2id, Version::V0x13, argon2id_params);
+
+        // Want at least 8 bytes salt, 16 bytes pw input.
+        if cleartext.len() < 24 {
+            return Err(TpmError::AuthValueTooShort);
+        }
```

Comment:
Checking min length should be the first thing we do IMO to short cut the whole process